### PR TITLE
Raven db 5213

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 1]
```diff
@@ -399,6 +400,96 @@ public Task GetImport()
...
+                            ["Type"] = "Error",
+                            ["Error"] = "Import from csv requires form content type"
+                        });
+                        return;
+                    }
+                }
+                var token = new OperationCancelToken(Database.DatabaseShutdown);
+                var result = new SmugglerResult();
+                var operationId = Database.Operations.GetNextOperationId();
+                await Database.Operations.AddOperation(Database, $"Import to: {Database.Name} from csv file", Raven.Server.Documents.Operations.Operations.OperationType.DatabaseImportFromCsv,
```
Comment (ml054) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: you await for operation and then result operation id, so this operation id is almost useless. 
Operation id should be taken from request uri - please have a look at import database endpoint. We also supply file and return operation id.

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 2]
```diff
@@ -399,6 +400,96 @@ public Task GetImport()
...
+            using (Database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+            {
+                if (HttpContext.Request.HasFormContentType == false)
+                {
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                    {
+                        context.Write(writer, new DynamicJsonValue
+                        {
+                            ["Type"] = "Error",
```
Comment (ml054) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:415]: nameof?

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 1 above)
Comment (ml054) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: I'm not databasename in task name is required. We already have this information in context.

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 3]
```diff
@@ -399,6 +400,96 @@ public Task GetImport()
...
+                                    if (MultipartRequestHelper.HasFileContentDisposition(contentDisposition))
+                                    {
+                                        if (ContentDispositionHeaderValue.TryParse(section.ContentDisposition, out contentDisposition) == false)
+                                            continue;
+
+                                        var filename = contentDisposition.FileName.ToString().Trim('\"');
+                                        var options = new DatabaseSmugglerOptionsServerSide();
+                                        if (section.Headers.ContainsKey("Content-Encoding") && section.Headers["Content-Encoding"] == "gzip")
+                                        {
+                                            using (var gzipStream = new GZipStream(section.Body, CompressionMode.Decompress))
```
Comment (ml054) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:451]: IIRC we have utility method from getting that?

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                {
+                    _hasId = true;
+                    _idIndex = i;
+                }
+                if (_csvReader.FieldHeaders[i].Equals(Constants.Documents.Metadata.Collection))
+                {
+                    _hasCollection = true;
+                    _collectionIndex = i;
+                }
+                if (_csvReader.FieldHeaders[i][0] == '@')
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:74]: Can we have empty field names?

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                    continue;
+                var indexOfDot = _csvReader.FieldHeaders[i].IndexOf('.');
+                //We have a nested property
+                if (indexOfDot >= 0)
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:78]: Field name `user.'name.first'` Note that there are quotes in the name. Needs to handle that in some manner?

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        private DocumentItem ConvertRecordToDocumentItem(string[] csvReaderCurrentRecord, string[] csvReaderFieldHeaders, string collection)
+        {
+            var idStr = _hasId ? csvReaderCurrentRecord[_idIndex] : _hasCollection? $"{csvReaderCurrentRecord[_collectionIndex]}/" : $"{collection}/";
+            var data = new DynamicJsonValue();
+            for (int i = 0; i < csvReaderFieldHeaders.Length; i++)
+            {
+                //ignoring reserved properties
+                if (csvReaderFieldHeaders[i][0] == '@')
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:112]: empty field name?

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: I think that the issue is the difference in uploading the file vs. processing the file. But yes, this should be split.

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            };
+        }
+
+        private unsafe object ParseValue(string s)
+        {
+            if (s.StartsWith('\"') && s.EndsWith('\"'))
+                return s.Substring(1, s.Length-2);
+            if (s.StartsWith('[') && s.EndsWith(']'))
+            {
+                var bytes = Encoding.UTF8.GetBytes(s);
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:169]: This allocates a temporary buffer per value, why? reuse an existing buffer

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                fixed (byte* ptr = bytes)
+                {
+                    return _context.ParseBufferToArray(ptr, bytes.Length, "CsvImport/ArrayValue", BlittableJsonDocumentBuilder.UsageMode.None);
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:172]: This is never disposed, which means that over time, this context is going to get VERY large.

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 6 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:112]: That's not valid csv

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: I have no idea what you guys mean here, this is exactly how it is done with smuggler import

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 3 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:451]: again exactly like in smuggler import

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 7 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:169]: Any suggestion would be appreciated

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 8 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:172]: I'll add disposable collection to DocumentItem and we will have to dispose of it once we are done processing it.

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 1 above)
Comment (ml054) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: ok you can await it but then operation id should be taken from request url, not generating here. https://github.com/ravendb/ravendb/blob/4b2c605ffa3af88a35ae66b828f5612414a8e3f6/src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs#L331

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 6 above)
Comment (ayende) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:112]: Like `Name,,Email` ? Note the double ,

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 2 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:415]: there is no property Type or Error in this context

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:424]: okay

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 3 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:451]: I don't know to what utility method you're referring to nor could i locate one even when you told me it exists, could you point me to where it reside?

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 4 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:74]: we can have an empty value but not an empty header

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 5 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:78]: done

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 6 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:112]: shouldn't happen in the headers line, can happen in the value lines

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 7 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:169]: done

File: src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs (hunk 8 above)
Comment (talweiss1982) [reply src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs:172]: done

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs (hunk 3 above)
Comment (ml054) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:451]: I meant something here: https://github.com/ravendb/ravendb/blob/4b2c605ffa3af88a35ae66b828f5612414a8e3f6/src/Raven.Server/Web/RequestHandler.cs#L77

**Pull request merged**